### ISSUE TITLE
Higher level model / training loop

### ISSUE THREAD

**Issue by kovasb** [google/jax#450]
Hi,

After writing some models by hand, I am missing having higher-level building blocks akin to Keras's Model or TF's Estimator/Head. 

Head is particularly interesting as a bite-sized building block that saves a decent amount of boiler plate without locking in too many assumptions. 

An imaginary API could look something like:

init_fn, train_step_fn = binary_classification_head(
    logits=stax.serial(...), 
    optimizer=optim.momentum(...),
    metrics={'metric_1': (metric_1_init_fn, metric_1_step_fn), ... }
)

initial_state = init_fn(input_shape)

trained = reduce(train_step_fn, initial_state, dataset_generator)

Where binary_classification_head composes all the init_fns and step_fns of its constituent components, and returns a single init_fn and train_step_fn. train_step_fn(state, batch) -> next_state

Nice thing is that init_fn, train_step_fn are subject to further JAX transforms, obviously jit but also vmap and others could be interesting. This is property that doesn't easily hold if one were to simply import designs from other frameworks without modification.

**Comment by kovasb** [comment]
Adding a breadcrumb here for other users who might be interested:

https://github.com/tensorflow/tensor2tensor/tree/master/tensor2tensor/trax

contains some interesting higher-level constructs and useful utility functions.

**Comment by mattjj** [comment]
Thanks for opening this (almost a year ago!) @kovasb.

This is out of scope for JAX core, but more higher-level libraries on top of JAX will be open-sourced in the very near future. Some of them include higher-level training loops.

I'm sure you'll notice when they get released (I'll be sure to retweet the announcement!).
